File: C++/347. Top K Frequent Elements.cpp
// 用优先队列， 小根堆， 超过k就pop
class Solution {
public:
    class myCompare {
    public:
        bool operator()(const pair<int, int>& a, const pair<int, int>& b) {
            return a.second > b.second;
        }
    };
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> map;
        for (int a : nums) map[a]++;
        priority_queue<pair<int, int>, vector<pair<int, int>>, myCompare> pri_que;
        for (auto m : map) {
            pri_que.push(m);
            if (pri_que.size() > k) pri_que.pop();
        }
        vector<int> res(k);
        for (int i = k - 1; i >= 0; i--) {
            res[i] = pri_que.top().first;
            pri_que.pop();
        }
        return res;

    }
};

File: Python/347. Top K Frequent Elements.py
# 用优先队列
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        map_ = {}
        for num in nums:
            map_[num] = map_.get(num, 0) + 1
        pri_que = []
        import heapq
        for key, freq in map_.items():
            heapq.heappush(pri_que, (freq, key))
            if len(pri_que) > k:
                heapq.heappop(pri_que)
        res = [0]*k
        for i in range(k-1, -1, -1):
            res[i] = heapq.heappop(pri_que)[1]
        return res